framework/decode/vulkan_virtual_swapchain.h

```diff
@@ -118,15 +106,32 @@ class VulkanVirtualSwapchain : public VulkanSwapchain
     {}
 
   private:
-    VkResult CreateSwapchainImage(const DeviceInfo*               device_info,
-                                  const VkImageCreateInfo&        image_create_info,
-                                  SwapchainKHRInfo::VirtualImage& image);
-
-    int32_t FindFirstPresentSrcLayout(const VkRenderPassCreateInfo* create_info) const;
-
-    int32_t FindFirstPresentSrcLayout(const VkRenderPassCreateInfo2* create_info) const;
-
-    int32_t FindFirstPresentSrcLayout(uint32_t count, const VkImageMemoryBarrier* barriers) const;
+    // VirtualSwapchain-specific private information.
+    struct VirtualImage
+    {
+        VkDeviceMemory                        memory{ VK_NULL_HANDLE };
+        VkImage                               image{ VK_NULL_HANDLE };
+        VulkanResourceAllocator::MemoryData   memory_allocator_data{ 0 };
+        VulkanResourceAllocator::ResourceData resource_allocator_data{ 0 };
+    };
+
+    struct PrivateData
```

Comment:
I think `PrivateData `should probably have a constructor which does all the work of creating these objects (and so may throw an exception or test failure some other way) and also have all the data it needs inside it to destroy itself (probably needs the device handle).  Then the `unordered_map `on `VkSwapchainKHR `should probably have `PrivateData` as the value and not a pointer, and the entry in the map should be emplaced with move semantics.  Then it correctly deletes itself when you call 

If that's not practical, `private_data_` should probably be an unordered_map with `unique_ptr<PrivateData>` values.

Also I suggest it should probably be named something more descriptive, like `SwapchainData`.

Reply:
I'll convert to a unique_ptr vector and rename it to SwapchainData.

Reply:
Done